From 67a4c301104a6f8bdb144ee9a33cbc431d7bbba7 Mon Sep 17 00:00:00 2001
From: srinivas pulukuru <srinivas.pulukuru@ti.com>
Date: Thu, 17 Mar 2011 17:38:07 -0500
Subject: OMAP4: DSS2: HDMI: Fix for HDMI notification to observer during HDMI_FIRST_HPD

if (hdmi_connected = false || hdmi_power != HDMI_POWER_FULL)
HDMI_FIRST_HPD should be ignored and no connect notification should be sent
to hdmi_obvserver. HDMI EDID read fails here and sending a notification
to hdmi_observer results in overlay interprepting the default VESA:VGA
timings as new panel timings and tries to configure the overlay accordingly.
This causes the overlay configuration to fail for TV manager if the
panel timings are smaller than the overlay then setmanager fails for TV
defaulting to LCD manager. This will cause both the overlays to be sent
to the LCD resulting in garbled display.

This happens only when trying to connect / disconnect the cable and
the connection is loose causing multiple connects / disconnects / first hpd
interrupts to be sent.

Change-Id: I2f9b876f697feaa730de0400461c6071f05f7e29
Signed-off-by: srinivas pulukuru <srinivas.pulukuru@ti.com>
---
 drivers/video/omap2/dss/hdmi.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/video/omap2/dss/hdmi.c b/drivers/video/omap2/dss/hdmi.c
index 869c58c..ccf8d51 100644
--- a/drivers/video/omap2/dss/hdmi.c
+++ b/drivers/video/omap2/dss/hdmi.c
@@ -1471,6 +1471,12 @@ static void hdmi_work_queue(struct work_struct *ws)
 		mutex_lock(&hdmi.lock);
 		mutex_lock(&hdmi.lock_aux);
 
+		if (hdmi_power != HDMI_POWER_FULL || !hdmi_connected) {
+			DSSINFO("irqstatus=0x%08x ignoring FIRST_HPD when "
+				"hdmi_connected = %d, hdmi_power = %d\n",
+				r, hdmi_connected, hdmi_power);
+			goto done;
+		}
 		/*
 		 * HDMI should already be full on. We use this to read EDID
 		 * the first time we enable HDMI via HPD.
-- 
1.6.0.4

